reports/glides_validator.livemd

```diff
@@ -22,90 +24,553 @@ Mix.install(
+    %{
+      name: "Mattapan",
+      tags: MapSet.new([:glides, :mattapan, :light_rail]),
+      stop_ids: MapSet.new(["70276"]),
+      # -> Capen Street
+      next: MapSet.new(["70274"])
+    }
+  ]
+
+  # %{stop_id => MapSet.t(stop_id)}
+  @first_to_next_stop @terminals
+                      |> Enum.flat_map(fn terminal ->
+                        Enum.map(terminal.stop_ids, &{&1, terminal.next})
+                      end)
+                      |> Map.new()
+
+  # %{stop_id => MapSet.t(stop_id)}
+  # (@first_to_next_stop, but inverted)
+  @next_to_first_stop @first_to_next_stop
+                      |> Enum.flat_map(fn {k, sets} -> Enum.map(sets, &{k, &1}) end)
+                      |> Enum.group_by(fn {_k, v} -> v end, fn {k, _v} -> k end)
+                      |> Map.new(fn {k, vs} -> {k, MapSet.new(vs)} end)
+
+  def first_to_next_stop, do: @first_to_next_stop
+
+  def next_to_first_stop, do: @next_to_first_stop
+
+  def by_tags(tags) when is_list(tags) do
+    by_tags(MapSet.new(tags))
+  end
+
+  def by_tags(tags) do
+    terminals()
+    |> Enum.filter(&MapSet.subset?(tags, &1.tags))
+    |> Enum.flat_map(& &1.stop_ids)
+    |> MapSet.new()
+  end
+
+  def by_name(name) do
+    Enum.find_value(terminals(), &if(&1.name == name, do: MapSet.new(&1.stop_ids)))
+  end
+
+  defp terminals, do: @terminals
+end
+```
+
+## Options for Report
+
 <!-- livebook:{"reevaluate_automatically":true} -->
 
 ```elixir
-date = Kino.Input.date("Date", default: Date.utc_today() |> Date.add(-1))
+env_input =
+  Kino.Input.select("Environment", [
+    {"", "prod"},
+    {"-dev-blue", "dev-blue"},
+    {"-dev-green", "dev-green"},
+    {"-dev", "dev"},
+    {"-sandbox", "sandbox"}
+  ])
+
+yesterday_eastern = DateTime.now!("America/New_York") |> DateTime.to_date() |> Date.add(-1)
+date_input = Kino.Input.date("Date", default: yesterday_eastern)
+
+stop_groups = [
+  {GlidesReport.Terminals.by_tags([:glides, :light_rail]), "All light rail terminal stops"},
+  {GlidesReport.Terminals.by_tags([:glides, :green]), "All Green Line terminal stops"},
+  {GlidesReport.Terminals.by_tags([:glides, :green, :western]),
+   "Western Green Line terminal stops"},
+  {GlidesReport.Terminals.by_tags([:glides, :green, :northern]),
+   "Northern Green Line terminal stops"}
+]
+
+stops =
+  [
+    "Boston College",
+    "Cleveland Circle",
+    "Riverside",
+    # (Not a Glides terminal)
+    # "Heath Street",
+    "Union Square",
+    "Medford/Tufts",
+    # (Not a Glides terminal)
+    # "Ashmont",
+    "Mattapan"
+  ]
+  |> Enum.map(&{GlidesReport.Terminals.by_name(&1), &1})
+
+stop_ids_input = Kino.Input.select("Stop(s)", stop_groups ++ stops)
+
+limit_to_next_2_predictions_input = Kino.Input.checkbox("Simulate countdown clocks?")
+
+sample_rate_input =
+  Kino.Input.range("Sample data at (?)-minute intervals", min: 1, max: 5, step: 1, default: 5)
+
+samples_per_minute_input =
+  Kino.Input.range("Take (?) samples per minute", min: 1, max: 10, step: 1, default: 1)
```

Comment:
One last non-blocking nit: I _think_ we output predictions every ~1s. So, at max, this means we are sampling 1/10th of out output RT files. Should we have a "DOWNLOAD ALL THE THINGS!!!!" option?

Reply:
I'd be worried about performance for that, both in terms of time and memory—I think the highest sample rate I tried was 5 samples every 1 minute, and it took _forever_ just to load all the data into memory, even on my shiny new M3 MBP with 18 GB of memory.

The two principal sources of slowness are:
- N+1 problem for S3 downloads: Each minute that we download data for requires a ListObjects query with prefix for that minute, a local `Enum.filter` through the results for objects we care about, and then a GetObject for each object we decide to download.
  - This problem may be partially mitigated by the file-naming changes I suggested [in slack](https://mbta.slack.com/archives/CF9QKK2AF/p1719501802477949?thread_ts=1719500671.500599&cid=CF9QKK2AF).
- Loading data into ETS tables: this is surprisingly slow.

203 seconds to download and store in memory 5760 files

Reply:
This is possible now! Cleaning up the data early turned out to really help with the size of the ETS tables. Download is still unavoidably slow (the first time) when this setting is active, since we have to download all trip update and vehicle position files for the day, but hopefully that's ok.

And once all the relevant data is downloaded, any subsequent reports generated for the same env + day are super snappy.